t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -1282,6 +1282,47 @@ test_expect_success 'sparse index is not expanded: read-tree' '
 	ensure_not_expanded read-tree --prefix=deep/deeper2 -u deepest
 '
 
+# NEEDSWORK: although the full repository's index is _not_ expanded as part of
+# stash, a temporary index, which is _not_ sparse, is created when stashing and
+# applying a stash of untracked files. As a result, the test reports that it
+# finds an instance of `ensure_full_index`, but it does not carry with it the
+# performance implications of expanding the full repository index.
```

Comment:
This is a good explanation of the problem. I wonder if there is something we can do to make creating this temporary index as a sparse one be better. Do you have a pointer to the place in the code that creates this index?

Reply:
It comes from setting the `GIT_INDEX_FILE` in `update-index` ([1](https://github.com/microsoft/git/blob/vfs-2.33.0/builtin/stash.c#L1089-L1092), [2](https://github.com/microsoft/git/blob/vfs-2.33.0/builtin/stash.c#L1217-L1221)) and `read-tree/checkout-index` ([1](https://github.com/microsoft/git/blob/vfs-2.33.0/builtin/stash.c#L342-L355), [2](https://github.com/microsoft/git/blob/vfs-2.33.0/builtin/stash.c#L1132-L1134)) to [temporary filename](https://github.com/microsoft/git/blob/vfs-2.33.0/builtin/stash.c#L1710-L1712) while either constructing the stash or applying it. 

When creating the stash (with untracked entries included) `update-index` will be invoked while the temporary index file does not exist, causing it to enter [this block in `do_read_index`](https://github.com/microsoft/git/blob/vfs-2.33.0/read-cache.c#L2276-L2277) - because it doesn't ever read the existing index, the `sparse_index` is never set and it's treated as a full index. I tried adding a check to set `sparse_index` based on the `sparse_index` repo setting and/or `command_requires_full_index`, but I hit some segfaults when testing, so it needed a closer look.

The more complicated case is on the unstashing side - `read-tree` also gets a non-existent index file, but it never actually reads in the index before writing out the result of `unpack_trees` (so what was added in `do_read_index` is never reached). In that case, the `sparse_index` flag is only ever set [if a sparse directory is written to it](https://github.com/microsoft/git/blob/vfs-2.33.0/read-cache.c#L110-L111), which won't happen when unpacking a tree without comparing to the index.

All that to say - I agree that there's more that can be done to preserve "sparseness" of the temp index. I think complications come up when dealing with untracked in files sparse directories (likely part of what caused the segfaults in my earlier testing), but I don't see it being impossible to handle.